File: Controllers/MasterDataController.js
const chalk = require("chalk");
const { Router } = require("express");
const { MasterDataRepository } = require("../Repository/MasterDataRepository");
const { MasterDataBusiness } = require("../Business/MasterDataBusiness");
const { DialogFlowService } = require("../Services/DialogFlowService");
class MasterDataController {
  /**
   * Returns a router instance representing the StatisticsController
   * @returns {Router}
   */
  static getRouter() {
    const router = Router();
    router.get("/data/departments", MasterDataController.GetDepartmentsAsync);
    router.post("/data/departments", MasterDataController.CreateDepartmentAsync);
    router.delete("/data/departments/:departmentId", MasterDataController.DeleteDepartmentAsync);
    router.get("/data/knowledge", MasterDataController.GetKnowledgeAsync);
    router.post("/data/knowledge", MasterDataController.CreateKnowledgeAsync);
    router.delete("/data/knowledge/:Term", MasterDataController.DeleteKnowledgeAsync);
    router.get("/scheme/:collection", MasterDataController.GetCollectionSchemeAsync);
    return router;
  }

  /**
   * Returns a list of registered departments
   * @param {Request} req
   * @param {Response} res
   */
  static async GetDepartmentsAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    const departments = await MasterDataRepository.GetAllData("departments");
    res.json(departments);
  }

  /**
   * Returns a list of known definitions
   * @param {Request} req
   * @param {Response} res
   */
  static async GetKnowledgeAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      const knowledgebase = await MasterDataBusiness.GetKeywordsAsync();
      res.json(knowledgebase);
    } catch (error) {
      console.log(chalk.yellow("Bad Request while getting knowledgebase:\n", error));
      res.status(400).send();
    }
  }

  /**
   * Returns the scheme for the specified collection
   * @param {Request} req
   * @param {Response} res
   */
  static async GetCollectionSchemeAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      const scheme = await MasterDataRepository.GetCollectionScheme(req.params.collection);
      res.json(scheme);
    } catch (error) {
      console.log(chalk.yellow("Bad Request while getting collection scheme:\n", error));
      res.status(400).send();
    }
  }

  /**
   * Creates a new knowledge
   * @param {Request} req
   * @param {Response} res
   */
  static async CreateKnowledgeAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      const result = await MasterDataBusiness.CreateKnowledgeAsync(req.body);

      if (result.error) {
        res.status(400).send(result);
        console.log(chalk.yellow("Bad request while creating knowledge:\n", JSON.stringify(result)));
        return;
      }

      const createdData = {
        name: req.body.name,
        keywords: [req.body.name, ...req.body.synonyms],
        definitiontype: req.body.definitiontype,
        description: req.body.description
      };
      res.json(createdData);
    } catch (error) {
      res.status(500).send();
      console.log(chalk.red("Unexpected error occured while creating knowledge:\n", error));
    }
  }

  /**
   * Deletes the knowledge about the specified Term
   * @param {Request} req
   * @param {Response} res
   */
  static async DeleteKnowledgeAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      const dialogFlowService = await DialogFlowService.getInstance();
      await Promise.all([
        MasterDataRepository.DeleteData("knowledge", { keyword: req.params.Term }),
        dialogFlowService.deleteKnowledge(req.params.Term)
      ]);

      res.status(204).send();
    } catch (error) {
      res.status(500).send();
      console.log(chalk.red("Unexpected error occured while deleting knowledge:\n", error));
    }
  }

  /**
   * Deletes a department
   * @param {Request} req
   * @param {Response} res
   */
  static async DeleteDepartmentAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      if (!req.params.departmentId) {
        res.status(400).send();
        return;
      }

      await MasterDataRepository.DeleteDataById("departments", req.params.departmentId);
      res.status(204).send();
    } catch (error) {
      res.status(500).send();
      console.log(chalk.red("Unexpected error occured while deleting department:\n", error));
    }
  }

  /**
   * Creates a department
   * @param {Request} req
   * @param {Response} res
   */
  static async CreateDepartmentAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      return;
    }

    try {
      if (typeof req.body.departmentName !== "string") {
        res.status(400).send();
        return;
      }

      const result = await MasterDataRepository.InsertCollectionData("departments", {
        departmentName: req.body.departmentName
      });
      const response = { ...req.body, _id: result.insertedId };
      res.send(response);
    } catch (error) {
      res.status(500).send();
      console.log(chalk.red("Unexpected error occured while creating department:\n", error));
    }
  }
}

module.exports = { MasterDataController };

File: src/ChatBot.Service/Controllers/MasterDataController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ChatBot.Business.Contracts.MasterData;
using ChatBot.Business.Contracts.MasterData.Exceptions;
using ChatBot.Business.Contracts.MasterData.Models;
using ChatBot.Business.Contracts.User;
using ChatBot.Repository.Contracts;
using ChatBot.Repository.Contracts.Models;
using ChatBot.Service.Mappers;
using ChatBot.Service.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Shouldly;

namespace ChatBot.Service.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class MasterDataController : ControllerBase
    {
        private readonly IDepartmentRepository _departmentRepository;
        private readonly ILogger<MasterDataController> _logger;
        private readonly IMasterDataBusiness _masterDataBusiness;
        private readonly IUserBusiness _userBusiness;

        public MasterDataController(IMasterDataBusiness masterDataBusiness, ILogger<MasterDataController> logger,
            IUserBusiness userBusiness, IDepartmentRepository departmentRepository)
        {
            masterDataBusiness.ShouldNotBeNull();
            userBusiness.ShouldNotBeNull();
            logger.ShouldNotBeNull();
            departmentRepository.ShouldNotBeNull();

            _masterDataBusiness = masterDataBusiness;
            _logger = logger;
            _userBusiness = userBusiness;
            _departmentRepository = departmentRepository;
        }

        /// <summary>
        ///     Creates a new department
        /// </summary>
        /// <param name="department"></param>
        /// <returns></returns>
        [Authorize]
        [HttpPost("data/departments")]
        [ProducesResponseType(typeof(DepartmentResponse), 200)]
        [ProducesResponseType(400)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreateDepartmentAsync([FromBody] CreateDepartmentRequest department)
        {
            try
            {
                _logger.LogInformation("Creating new department");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate department data.");
                    return Unauthorized();
                }

                var newUser = await _masterDataBusiness
                    .CreateDepartmentAsync(department.DepartmentName)
                    .ConfigureAwait(false);

                return Ok(newUser.Map());
            }
            catch (DepartmentAlreadyExistsException)
            {
                _logger.LogInformation("Could not create department because it already exists");
                return BadRequest();
            }
            catch (MissingDataException)
            {
                _logger.LogInformation("Could not create department due to missing data");
                return BadRequest();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error while creating new user");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns all registered departments
        /// </summary>
        /// <returns></returns>
        [Authorize]
        [HttpGet("data/departments")]
        [ProducesResponseType(typeof(IEnumerable<Department>), 200)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> GetAllDepartmentsAsync()
        {
            try
            {
                _logger.LogInformation("Loading all departments");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate department data.");
                    return Unauthorized();
                }

                var departments = await _departmentRepository
                    .GetAllDepartmentsAsync()
                    .ConfigureAwait(false);
                var response = departments.Select(RepositoryMapper.Map);

                return Ok(response);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Could not load all departments");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns the schema for the specified model
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        [Authorize]
        [HttpGet("scheme/{type}")]
        [ProducesResponseType(typeof(MasterDataSchema), 200)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> GetSchemaAsync([FromRoute] string type)
        {
            try
            {
                _logger.LogInformation("Trying to get scheme for master data type");
                var requestedType = ParseMasterDataType(type);

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to get data illegally.");
                    return Unauthorized();
                }

                var schema = await _masterDataBusiness.GetSchema(requestedType);
                return Ok(schema.Map());
            }
            catch (ArgumentOutOfRangeException)
            {
                _logger.LogInformation("Got invalid master data type");
                return BadRequest();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Could not load schema");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Deletes the specified department
        /// </summary>
        /// <param name="departmentId"></param>
        /// <returns></returns>
        [Authorize]
        [HttpDelete("data/departments/{departmentId}")]
        [ProducesResponseType(204)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> DeleteDepartmentAsync([FromRoute] string departmentId)
        {
            try
            {
                _logger.LogInformation($"Deleting department with id {departmentId}");

                if (string.IsNullOrWhiteSpace(departmentId))
                {
                    _logger.LogInformation("DepartmentId was invalid");
                    return NotFound();
                }

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate department data.");
                    return Unauthorized();
                }

                await _departmentRepository
                    .DeleteDepartmentAsync(departmentId)
                    .ConfigureAwait(false);

                return NoContent();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error occured while deleting a user");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns all known terms with explanations
        /// </summary>
        /// <returns></returns>
        [Authorize]
        [HttpGet("data/knowledge")]
        [ProducesResponseType(typeof(IEnumerable<KnowledgeResponse>), 200)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> GetKnowledgeAsync()
        {
            try
            {
                _logger.LogInformation("Loading knowledge base");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate knowledge data.");
                    return Unauthorized();
                }

                var knowledgeBase = await _masterDataBusiness
                    .GetKnowledgeBaseAsync()
                    .ConfigureAwait(false);

                return Ok(knowledgeBase.Select(knowledge => knowledge.Map()));
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error occured while getting knowledge base");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Deletes a term from the knowledge base
        /// </summary>
        /// <param name="definitionType"></param>
        /// <param name="keyword"></param>
        /// <returns></returns>
        [Authorize]
        [HttpDelete("data/knowledge/{definitionType}/{keyword}")]
        [ProducesResponseType(201)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> DeleteKnowledgeAsync(string definitionType, string keyword)
        {
            try
            {
                _logger.LogInformation($"Trying to delete term ({definitionType} / {keyword})");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate knowledge data.");
                    return Unauthorized();
                }

                await _masterDataBusiness.DeleteTermAsync(definitionType, keyword);
                return NoContent();
            }
            catch (MissingDataException ex)
            {
                _logger.LogWarning(ex, "The user provided insufficient information");
                return BadRequest();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error occured while deleting term");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Creates the specified knowledge
        /// </summary>
        /// <param name="knowledge"></param>
        /// <returns></returns>
        [Authorize]
        [HttpPost("data/knowledge")]
        [ProducesResponseType(typeof(KnowledgeResponse), 200)]
        [ProducesResponseType(400)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreateKnowledgeAsync([FromBody] CreateKnowledgeRequest knowledge)
        {
            try
            {
                _logger.LogInformation("Trying to create knowledge");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate knowledge data.");
                    return Unauthorized();
                }

                var newKnowledge = await _masterDataBusiness
                    .CreateKnowledgeAsync(knowledge.Map())
                    .ConfigureAwait(false);

                return Ok(newKnowledge.Map());
            }
            catch (MissingDataException ex)
            {
                _logger.LogWarning(ex, "The user provided insufficient information");
                return BadRequest();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error occured while deleting term");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns the id of the current user id
        /// </summary>
        /// <returns></returns>
        private string GetCurrentUserId()
        {
            var userId = User.Claims.FirstOrDefault(claim => claim.Type == "user_id")?.Value;
            userId.ShouldNotBeNullOrWhiteSpace();
            return userId;
        }

        /// <summary>
        ///     Converts a string to a master data type
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        private static MasterDataType ParseMasterDataType(string type)
        {
            if (type.Equals("knowledge", StringComparison.InvariantCultureIgnoreCase))
                return MasterDataType.Knowledge;
            if (type.Equals("departments", StringComparison.InvariantCultureIgnoreCase))
                return MasterDataType.Department;
            throw new ArgumentOutOfRangeException(nameof(type));
        }
    }
}